File: Irrgarten-Java/src/irrgarten/Labyrinth.java
/**
 * @file irrgarten/Labyrinth.java
 * @author Álvaro Maldonado Medina, grupo 2D-D3
 */

package irrgarten;

import static irrgarten.Directions.DOWN;
import static irrgarten.Directions.LEFT;
import static irrgarten.Directions.RIGHT;
import static irrgarten.Directions.UP;
import java.util.ArrayList;

/**
 * Clase encargada de generar el esqueleto del juego, el laberinto donde se va
 * a desarrollar toda la acción, y de gestionarlo. Tiene forma de matriz 2D.
 */

public class Labyrinth {
    
    //*****************************************************************
    // Constantes 
    //*****************************************************************
    
    private static final char BLOCK_CHAR = 'X'; // indicador de bloques
    private static final char EMPTY_CHAR = '-'; // indicador de casilla vacía
    private static final char MONSTER_CHAR = 'M';   // indicador de monstruo
    private static final char COMBAT_CHAR = 'C';    // indicador de casilla de combate
    private static final char EXIT_CHAR = 'E';  // indicador de casilla de salida
    private static final int ROW = 0; 
    private static final int COL = 1;     
    private int nRows, nCols, exitRow, exitCol; // nº de filas y columnas, fila 
                                                // y columna de la casilla de salida
    
    public Monster[][] monsters; // monstruos en el laberinto
    public Player[][] players;  // jugadores en el laberinto
    public char[][] labyrinth;  // laberinto
    
    
    /**
     * Constructor del laberinto
     * @param nRows Número de filas
     * @param nCols Número de columnas
     * @param exitRow Número de fila de la casilla de salida
     * @param exitCol Número de columna de la casilla de salida
     */
       
    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
     
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        
        this.monsters = new Monster[nRows+1][nCols+1];
        this.players = new Player[nRows+1][nCols+1];
        this.labyrinth = new char[nRows+1][nCols+1];
        
        for (int i = 0; i < nRows; i++){
            for (int j = 0; j < nCols; j++){
                labyrinth[i][j] = EMPTY_CHAR;
            }
        }
        
        labyrinth[exitRow][exitCol] = EXIT_CHAR;
        
    }

    
    /**
     * Método que se encarga de calcular si hay un ganador o no.
     * @return true si hay un jugador en la casilla de salida.
     */
    
    public boolean haveAWinner()
    {
        return players[exitRow][exitCol] != null;
    }
    
    /**
     * Método toString
     * @return Cadena de caracteres con la matriz del laberinto y su estado actual
     */
    
    @Override
    public String toString()
    {
        String estado = "";
        
        for (int i=0; i < nRows; i++){
            for (int j=0; j < nCols; j++){
                estado += " ["+labyrinth[i][j]+"]";
            }
            estado += "\n";
        }
        
        estado +="\n";
        estado +="Tamaño del laberinto: "+getRows()+"x"+getCols()+"\n"
                + "Casilla de salida: ["+exitRow+","+exitCol+"]";
        return estado;
        
        
    }
    
    /**
     * Calcula si la posición pasada como parámetro es correcta o no
     * @param row Fila
     * @param col Columna
     * @return true si la posición es correcta
     */
    
    private boolean posOK(int row, int col){
        return (row >= 0) && (row <= nRows) && (col >= 0) && (col <= nCols);
    }
    
   /**
    * Añade monstruos al laberinto, y al contenedor de éstos
    * @param row Fila de la casilla donde se añade el monstruo
    * @param col Columna de la casilla donde se añade el monstruo
    * @param monster Monstruo
    */
    
    public void addMonster(int row, int col, Monster monster)
    {           
        while(!emptyPos(row,col)){
            row = Dice.randomPos(getRows());
            col = Dice.randomPos(getCols());
        }
        
        labyrinth[row][col] = MONSTER_CHAR;
        monsters[row][col] = monster;
        monster.setPos(row, col);
    }
    
    /**
     * Método que se encarga de esparcir a los jugadores por las posiciones
     * válidas del laberinto que quedan libres tras la configuración
     * @param players Lista de jugadores
     */
    
    public void spreadPlayers(ArrayList<Player> players)
    {
        for (int i = 0; i < players.size(); i++)
        {
            Player p = players.get(i);
            ArrayList<Integer> pos = randomEmptyPos();            
            putPlayer2D(-1,-1, pos.get(ROW), pos.get(COL), p);     
        }
    }
    
    /**
     * Calcula si la posición pasada como parámetro está o no vacía
     * @param row Fila
     * @param col Columna
     * @return true si la posición está vacía
     */
    
    private boolean emptyPos(int row, int col)
    {           
        return labyrinth[row][col] == EMPTY_CHAR;
    }
    
    /**
     * Método que indica si una posición pasada como parámetro 
     * contiene un monstruo o no
     * @param row Fila de la casilla
     * @param col Columna de la casilla
     * @return True o false dependiendo de si es una monsterPos o no
     */
    
    private boolean monsterPos(int row, int col)
    {        
        return labyrinth[row][col] == MONSTER_CHAR;
    }
    
    /**
     * Calcula si la posición dada como parámetro es la de salida
     * @param row Fila de la casilla
     * @param col Columna de la casilla
     * @return True o false dependiendo de si es o no la exitPos
     */
    
    private boolean exitPos(int row, int col)
    {
        return labyrinth[row][col] == EXIT_CHAR;
    }
    
    /**
     * Calcula si la posicón dada como parámetro es de combate o no
     * @param row Fila
     * @param col Columna
     * @return True o false dependiendo de si es combatPos
     */
    
    private boolean combatPos(int row, int col)
    {        
        return labyrinth[row][col] == COMBAT_CHAR;
    }
    
    /**
     * Indica si en una posición el jugador puede moverse sobre ella o no
     * @param row Fila
     * @param col Columna
     * @return True o false si puede caminar o no sobre la posición
     */
    
    private boolean canStepOn(int row, int col)
    {           
        if (posOK(row,col))
        {
            if (emptyPos(row, col) || monsterPos(row, col) || exitPos(row, col))
            {
                return true;
            }
        
        }
        return false;
    }
    
    /**
     * Actualiza una posición a su estado anterior a que el jugador pasara
     * sobre ella: si era de combate, pasa a ser de monstruo, o de lo contrario
     * pasa a ser casilla vacía
     * @param row Fila
     * @param col Columna
     */
    
    private void updateOldPos(int row, int col)
    {
        if (posOK(row, col))
        {
            if (combatPos(row, col))
                labyrinth[row][col] = MONSTER_CHAR;
            else
                labyrinth[row][col] = EMPTY_CHAR;
        }
    }
    
    /**
     * Determina la posición a la que se va a mover un jugador en función de 
     * la dirección que tome
     * @param row Fila
     * @param col Columna
     * @param direction Dirección de movimiento
     * @return La nueva posición a la que se moverá el jugador
     */
    
    private ArrayList<Integer> dir2Pos(int row, int col, Directions direction)
    {
        if (posOK(row,col)){
            switch (direction)
            {
                case UP:
                    row--;
                    break;
                case DOWN:
                    row++;
                    break;
                case LEFT:
                    col--;
                    break;
                case RIGHT:
                    col++;
                    break;
            }
        }
        
        ArrayList<Integer> newPos = new ArrayList<>();
        newPos.add(row);
        newPos.add(col);
        
        return newPos;
    }
    
    /**
     * Genera posiciones aleatorias usando el método randomPos de la clase Dice
     * hasta dar con una que esté vacía. Mientras tanto, se mantiene en un bucle
     * infinito
     * @return Una posición válida generada aleatoriamente
     */
    
    private ArrayList<Integer> randomEmptyPos()
    {
        ArrayList<Integer> randomPos= new ArrayList<>();
        
        int randomRow;
        int randomCol;
        
        do{
            randomRow = Dice.randomPos(nRows);
            randomCol = Dice.randomPos(nCols);
        } while(!(emptyPos(randomRow,randomCol)));
        
        randomPos.add(randomRow);
        randomPos.add(randomCol);
        
        return randomPos;
    }
    
    /**
     * Mueve un jugador a una posición determinada y devuelve un posible monstruo
     * si el movimiento resulta en la presencia de uno en la nueva posición del
     * jugador
     * @param direction Dirección de movimiento del jugador
     * @param player Jugador
     * @return El posible monstruo.
     */
    
    public Monster putPlayer(Directions direction, Player player)
    {
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        
        ArrayList<Integer> newPos = dir2Pos(oldRow, oldCol, direction);
        Monster monster = putPlayer2D(oldRow, oldCol, newPos.get(ROW), newPos.get(COL), player);
        
        return monster;
    }
    
    /**
     * Método que gestiona la creación y adición de un bloque o una fila
     * de éstos en hileras verticales u horizontales en el laberinto
     * @param orientation Orientación de la hilera de bloques
     * @param startRow Fila de inicio de la hilera
     * @param startCol Columna de inicio de la hilera
     * @param length Longitud de la hilera
     */
    
    public void addBlock(Orientation orientation, int startRow, int startCol,
            int length)
    {
        int incRow, incCol;
        
        if (orientation == Orientation.VERTICAL)
        {
             incRow = 1;
             incCol = 0;
        } else {
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow, col = startCol;
        
        while ((posOK(row, col)) && (emptyPos(row,col)) && (length > 0))
        {
            labyrinth[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }
    
    /**
     * Método que crea una lista con todos los posibles movimientos válidos que
     * puede tomar un jugador antes de que éste introduzca una dirección de
     * movimiento por teclado
     * @param row Fila
     * @param col Columna
     * @return La lista de movimientos válidos mencionada
     */
    
    public ArrayList<Directions> validMoves(int row, int col)
    {
        ArrayList<Directions> output = new ArrayList <>();
        
        if(canStepOn(row+1,col)){
            output.add(Directions.DOWN);
        }
        if(canStepOn(row-1,col)){
            output.add(Directions.UP);
        }
        if(canStepOn(row,col+1)){
            output.add(Directions.RIGHT);
        }
        if(canStepOn(row,col-1)){
            output.add(Directions.LEFT);
        }
        
        return output;
    }
    
    /**
     * Método que coloca jugadores en posiciones bidimensionales en el laberinto
     * y actualiza el estado de la posición en función de donde éste haya caído.
     * @param oldRow Fila de la casilla anterior del jugador
     * @param oldCol Estado anterior de la columna de la casilla
     * @param row Fila de la nueva casilla
     * @param col Columna de la nueva casilla
     * @param player Jugador
     * @return un output en forma de monstruo, nulo si no lo había, monstruo si lo había
     */
    
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, 
            Player player)
    {
        
        Monster output = null;
        
        if (canStepOn(row, col))
        {
            if (posOK(oldRow, oldCol))
            {
                Player p = players[oldRow][oldCol];
                
                if (p == player)
                {
                    updateOldPos(oldRow,oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            
            boolean monsterPos = monsterPos(row, col);
            
            if (monsterPos)
            {
                labyrinth[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            } else {
                char number = player.getNumber();
                labyrinth[row][col] = number;
            }
            
            players[row][col] = player;
            player.setPos(row, col);
            
        }
        
        return output;
    }
    
    //*****************************************************************
    // Getters 
    //*****************************************************************
    
    /**
     * Row getter
     * @return Número de filas del laberinto
     */
    
    public int getRows()
    {
        return nRows;
    }
    
    /**
     * Column getter
     * @return Número de columnas del laberinto
     */
    
    public int getCols()
    {
        return nCols;
    }
    
    /**
     * Coloca un FuzzyPlayer en el mismo sitio donde muere un Player, al resucitar
     * NOTA: al resucitar, el FuzzyPlayer no aparece, pero en el momento en que se mueva
     * a una nueva casilla, éste aparecerá sin problema. He tratado de arreglarlo
     * de varias maneras y no funciona
     * 
     * @param fuzzyPlayer El FuzzyPlayer
     * @param row Fila de la casilla donde se coloca la instancia
     * @param col Columna de la casilla donde se coloca la instancia
     */
    
    public void placeFuzzyPlayer(FuzzyPlayer fuzzyPlayer, int row, int col){
        if (posOK(row,col)){
            labyrinth[row][col] = EMPTY_CHAR;
            fuzzyPlayer.setPos(row, col);
            
        }
    }
      
    
}

File: Irrgarten-Ruby/irrgarten/labyrinth.rb
#encoding:utf-8

#######################################
# Álvaro Maldonado Medina             #
# 2D- D3                              #
# Fichero: Irrgarten/Labyrinth.rb     #
#######################################

require 'matrix'
require_relative 'monster'
require_relative 'player'
require_relative 'dice'
require_relative 'shield'
require_relative 'weapon'
require_relative 'directions'
require_relative 'orientation'
require_relative 'fuzzy_player'

module Irrgarten

  class Labyrinth

	 attr_reader :n_rows, :n_cols

	 @@BLOCK_CHAR = 'X'
	 @@EMPTY_CHAR = '-'
	 @@MONSTER_CHAR = 'M'
	 @@COMBAT_CHAR = 'C'
	 @@EXIT_CHAR = 'E'
	 @@ROW = 0
	 @@COL = 1

	 # Constructor
	 def initialize (n_rows, n_cols, exit_row, exit_col)

		@n_rows = n_rows
		@n_cols = n_cols
		@exit_row = exit_row
		@exit_col = exit_col

		@monsters = Array.new(n_rows+1) { Array.new(n_cols+1) }
		@players = Array.new(n_rows+1) { Array.new(n_cols+1) }
		@labyrinth = Array.new(n_rows+1) { Array.new(n_cols+1, @@EMPTY_CHAR) }
		@labyrinth[exit_row][exit_col] = @@EXIT_CHAR
	 end

	 # Método encargado de esparcir los jugadores por el mapa al comenzar la partida
	 def spread_players(players)
		players.each do |i|
		  pos = random_empty_pos
		  put_player_2D(-1,-1,pos[@@ROW],pos[@@COL],i)
		end
	 end

	 # Método encargado de chivar si hay un ganador
	 def have_a_winner
		@players[@exit_row][@exit_col] != nil
	 end

	 # To_String
	 def to_s
		estado = ""

		(0...@n_rows).each do |i|
		  (0...@n_cols).each do |j|
			 estado += " [#{@labyrinth[i][j]}]"
		  end
		  estado += "\n"
		end

		estado += "===============================================================\n"
		estado += "Tamaño del laberinto: #{@n_rows}x#{@n_cols}\n"
		estado += "Casilla de salida: [#{@exit_row},#{@exit_col}]\n"
		estado += "===============================================================\n"

		estado
	 end

	 # Método encargado de añadir y setear un monstruo en una posición del laberinto
	 def add_monster(row, col, monster)

		if pos_ok(row,col)
		  if empty_pos(row,col)

			 @labyrinth[row][col] = @@MONSTER_CHAR
			 @monsters[row][col] = monster

			 if (@labyrinth[row][col] != @@BLOCK_CHAR) && (@labyrinth[row][col] != @@EXIT_CHAR)
				monster.set_pos(row,col)
			 end

		  end
		end

	 end

	 # Método encargado de mover a un jugador y devolver un posible monstruo si la posición es M
	 def put_player(direction, player)
		old_row = player.row
		old_col = player.col

		new_pos = dir_2_pos(old_row, old_col, direction)
		put_player_2D(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
	 end

	 # Añade muros bloques al laberinto de diferente longitud y orientación
	 def add_block(orientation, start_row, start_col, length)

		if orientation == Orientation::VERTICAL
		  inc_row = 1
		  inc_col = 0
		else
		  inc_row = 0
		  inc_col = 1
		end

		row = start_row
		col = start_col

		while pos_ok(row,col) && empty_pos(row,col) && length > 0
		  @labyrinth[row][col] = @@BLOCK_CHAR
		  length -= 1
		  row += inc_row
		  col += inc_col
		end

	 end

	 # Calcula las direcciones válidas posibles de movimiento para el jugador en el mapa
	 def valid_moves(row, col)
		output = Array.new

		if can_step_on(row+1,col)
		  output.push(Directions::DOWN)
		end
		if can_step_on(row-1,col)
		  output.push(Directions::UP)
		end
		if can_step_on(row,col+1)
		  output.push(Directions::RIGHT)
		end
		if can_step_on(row,col-1)
		  output.push(Directions::LEFT)
		end

		output
	 end

	 # Getter de la fila
	 def rows
		@n_rows
	 end

	 # Getter de la columna
	 def cols
		@n_cols
	 end

	 private
	 # Calcula si la posición del laberinto pasada como parámetro es correcta
	 def pos_ok(row, col)
		(row >= 0) && (row <= @n_rows) && (col >= 0) && (col <= @n_cols)
	 end

	 # Devuelve si la posición pasada como parámetro es vacía
	 def empty_pos(row, col)
		@labyrinth[row][col] == @@EMPTY_CHAR
	 end

	 # Devuelve si la posición pasada como parámetro es de monstruo
	 def monster_pos(row, col)
		@labyrinth[row][col] == @@MONSTER_CHAR
	 end

	 # Devuelve si la posición pasada como parámetro es la de salida
	 def exit_pos(row, col)
		@labyrinth[row][col] == @@EXIT_CHAR
	 end

	 # Devuelve si la posición pasada como parámetro es de combate
	 def combat_pos(row, col)
		@labyrinth[row][col] == @@COMBAT_CHAR
	 end

	 # Devuelve si es posible moverse sobre la posición pasada como parámetro
	 def can_step_on(row, col)
		if @labyrinth[row][col] != @@BLOCK_CHAR
		  pos_ok(row,col) and ((empty_pos(row,col) or monster_pos(row,col) or exit_pos(row, col)))
		end
	 end

	 # Actualiza la posición del laberinto a su estado anterior previo a un combate o un movimiento del jugador sobre ella
	 def update_old_pos(row, col)

		if pos_ok(row,col)
		  if @labyrinth[row][col] == @@COMBAT_CHAR
			 @labyrinth[row][col] = @@MONSTER_CHAR
		  else
			 @labyrinth[row][col] = @@EMPTY_CHAR
		  end
		end
	 end

	 # Calcula la dirección hacia una posición determinada
	 def dir_2_pos(row, col, direction)

		new_row=row
		new_col=col

		case direction
		when Directions::UP
		  new_row=new_row-1
		when Directions::DOWN
		  new_row=new_row+1
		when Directions::LEFT
		  new_col=new_col-1
		when Directions::RIGHT
		  new_col=new_col+1
		end
		[new_row, new_col]

	 end

	 # Calcula posiciones aleatorias hasta que sale una vacía
	 def random_empty_pos
		resultado = Array.new

		row = Dice.random_pos(@n_rows)
		col = Dice.random_pos(@n_cols)

		loop do
		  if pos_ok(row,col)
			 resultado[@@ROW] = row
			 resultado[@@COL] = col
		  end

		  if empty_pos(resultado[@@ROW], resultado[@@COL]) && @labyrinth[@@ROW, @@COL] != @@BLOCK_CHAR
			 break
		  end
		end

		resultado
	 end

	 # Coloca jugadores en posiciones del laberinto y actualiza el estado de la posición en función de donde haya caído este
	 def put_player_2D(old_row, old_col, row, col, player)

		output = nil

		if can_step_on(row, col)
		  if pos_ok(row, col)
			 p = @players[old_row][old_col]

			 if p == player
				update_old_pos(old_row, old_col)
				@players[old_row][old_col] = nil
			 end

		  end

		  monster_pos = monster_pos(row, col)
		  if monster_pos
			 @labyrinth[row][col] = @@COMBAT_CHAR
			 output = @monsters[row][col]
		  else
			 number = player.number
			 @labyrinth[row][col] = number
		  end

		  @players[row][col] = player
		  player.set_pos(row, col)

		end

		output

	 end

	 public
	 # Nº de filas del laberinto
	 def n_rows
		@n_rows
	 end

	 # Nº de columnas del laberinto
	 def n_cols
		@n_cols
	 end

	 # Método creado para setear la posición de un fuzzy_player
	 def place_fuzzy_player(fuzzy_player, row, col)
		if pos_ok(row,col)
		  @labyrinth[row][col] = @@EMPTY_CHAR
		  fuzzy_player.set_pos(row,col)
		end
	 end

  end	# class
end	# module